engine/lib/phases/phase_project_instead_of_match.ml

```diff
@@ -0,0 +1,227 @@
+    include F
+    include Features.Off.Project_instead_of_match
+  end
+
+  include
+    Phase_utils.MakeBase (F) (FB)
+      (struct
+        let phase_id = Diagnostics.Phase.ProjectInsteadOfMatch
+      end)
+
+  module UA = Ast_utils.Make (F)
+  module UB = Ast_utils.Make (FB)
+
+  module Implem : ImplemT.T = struct
+    let metadata = metadata
+
+    module S = struct
+      include Features.SUBTYPE.Id
+      include Features.Off.Project_instead_of_match
+    end
+
+    [%%inline_defs dmutability]
+
+    let rec dty (span : span) (ty : A.ty) : B.ty =
+      match ty with [%inline_arms "dty.*"] -> auto
+
+    and dpat' (span : span) (pat : A.pat') : B.pat' =
+      match pat with
+      | [%inline_arms "dpat'.*" - PConstruct] -> auto
+      | PConstruct { name; args; is_record = _; is_struct } ->
+          PConstruct
+            {
+              name;
+              args = List.map ~f:(dfield_pat span) args;
+              is_record = None;
+              is_struct;
+            }
+
+    and project_pat (p : A.pat) : B.pat * (B.pat * B.expr) list =
+      let simple_pat, remaining_pats = project_pat' p.span p.p in
+      ({ p = simple_pat; span = p.span; typ = dty p.span p.typ }, remaining_pats)
+
+    and project_field_pat (_span : span) (p : A.field_pat) :
+        B.field_pat * (B.pat * B.expr) list =
+      let pat, pat_list = project_pat p.pat in
+      ({ field = p.field; pat }, pat_list)
+
+    and project_pat' (span : span) (pat : A.pat') :
+        B.pat' * (B.pat * B.expr) list =
+      match pat with
+      | PWild -> (PWild, [])
+      | PAscription { typ; typ_span; pat } ->
+          let simple_pat, remaining_pats = project_pat pat in
+          ( PAscription { typ = dty span typ; pat = simple_pat; typ_span },
+            remaining_pats )
+      | PConstruct { name; args; is_record = Some _; is_struct } ->
+          let update_args = List.map ~f:(project_field_pat span) args in
+          let new_id = UA.fresh_local_ident_in_expr [] "some_name" in
+          ( PConstruct
+              {
+                name;
+                args =
+                  [
+                    {
+                      field = name;
+                      pat =
+                        {
+                          p =
+                            PBinding
+                              {
+                                mut = Immutable;
+                                mode = ByValue;
+                                var = new_id (* name *);
+                                typ = TApp { ident = name; args = [] };
+                                (* TODO? *)
+                                subpat = None;
+                              };
+                          span;
+                          typ = TApp { ident = name; args = [] };
+                        };
+                    };
+                  ];
+                is_record = (None : FB.project_instead_of_match option);
+                is_struct;
+              },
```

Comment:
Should that actually be a `PConstruct`? Here we rewrite `Foo {a: ..., b: ..., c: ...}` into `Foo {Foo: some_name}`, that's wrong right?

Reply:
I feel like this should be rather be a `PBinding`